internal/blocksync/reactor.go

```diff
@@ -489,94 +485,112 @@ func (r *Reactor) poolRoutine(ctx context.Context, stateSynced bool, blockSyncCh
-				r.logger.Error(
-					err.Error(),
-					"last_commit", second.LastCommit,
-					"block_id", firstID,
-					"height", first.Height,
-				)
+			if r.lastTrustedBlock != nil {
+				err := VerifyNextBlock(newBlock, newBlockID, verifyBlock, r.lastTrustedBlock.block, r.lastTrustedBlock.commit, state.NextValidators)
+
+				if err != nil {
+					switch err.(type) {
+					case ErrBlockIDDiff:
+					case ErrValidationFailed:
+						peerID := r.pool.RedoRequest(r.lastTrustedBlock.block.Height + 1)
+						if serr := blockSyncCh.SendError(ctx, p2p.PeerError{
+							NodeID: peerID,
+							Err:    errors.New("invalid block for verification"),
+						}); serr != nil {
+							return
+						}
+					case ErrInvalidVerifyBlock:
+						r.logger.Error(
+							err.Error(),
+							"last_commit", verifyBlock.LastCommit,
+							"block_id", newBlockID,
+							"height", r.lastTrustedBlock.block.Height,
+						)
+						peerID := r.pool.RedoRequest(r.lastTrustedBlock.block.Height + 2)
+						if serr := blockSyncCh.SendError(ctx, p2p.PeerError{
+							NodeID: peerID,
+							Err:    err,
+						}); serr != nil {
+							return
+						}
+					}
 
-				// NOTE: We've already removed the peer's request, but we still need
-				// to clean up the rest.
-				peerID := r.pool.RedoRequest(first.Height)
-				if serr := blockSyncCh.SendError(ctx, p2p.PeerError{
-					NodeID: peerID,
-					Err:    err,
-				}); serr != nil {
-					return
 				}
 
-				peerID2 := r.pool.RedoRequest(second.Height)
-				if peerID2 != peerID {
-					if serr := blockSyncCh.SendError(ctx, p2p.PeerError{
-						NodeID: peerID2,
-						Err:    err,
-					}); serr != nil {
-						return
+			} else {
+				// we need to load the last block we trusted
+				if r.initialState.LastBlockHeight != 0 {
+					r.lastTrustedBlock = &BlockResponse{r.store.LoadBlock(r.initialState.LastBlockHeight), r.store.LoadBlockCommit(r.initialState.LastBlockHeight)}
+					if r.lastTrustedBlock == nil {
+						panic("Failed to load last trusted block")
 					}
 				}
-			} else {
-				r.pool.PopRequest()
-
-				// TODO: batch saves so we do not persist to disk every block
-				r.store.SaveBlock(first, firstParts, second.LastCommit)
+				oldHash := r.initialState.Validators.Hash()
+				if !bytes.Equal(oldHash, newBlock.ValidatorsHash) {
 
-				var err error
+					fmt.Println(
 
-				// TODO: Same thing for app - but we would need a way to get the hash
-				// without persisting the state.
-				state, err = r.blockExec.ApplyBlock(ctx, state, firstID, first)
-				if err != nil {
-					// TODO: This is bad, are we zombie?
-					panic(fmt.Sprintf("failed to process committed block (%d:%X): %v", first.Height, first.Hash(), err))
+						"initial hash ", r.initialState.Validators.Hash(),
+						"new hash ", newBlock.ValidatorsHash,
+					)
+					return
 				}
+				r.lastTrustedBlock = &BlockResponse{block: newBlock, commit: verifyBlock.LastCommit}
+			}
+			r.lastTrustedBlock.block = newBlock
```

Comment:
But, if we came from line 559 (i.e., `else` in line 541), the blocks haven't been verified... am I missing something?

Reply:
Good catch, result of copy pasting..basically we need to contiunue the loop whenever verification fails and keep waiting for a new block!

Reply:
This is why adding a test with a scenario that should not validate is important. It would have caught all these cases.